src/rust/catmem/ring/pop_ring.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+//======================================================================================================================
+// Imports
+//======================================================================================================================
+
+use crate::{
+    collections::shared_ring::SharedRingBuffer,
+    runtime::fail::Fail,
+};
+
+//======================================================================================================================
+// Structures
+//======================================================================================================================
+
+/// An endpoint for a unidirectional queue built on a shared ring buffer
+/// The state of the underlying ring reference
+#[derive(Copy, Clone, Debug, PartialEq)]
+enum RingState {
+    /// A pipe that is opened.
+    Opened,
+    /// A pipe that is closed. For a consumer (pop) pipe, this means that it has received an EOF.
+    Closed,
+}
```

Comment:
This can be moved to a sibling module and be shared with PushRing.

Reply:
Since the states are not exactly identical, per Anand's suggestion, we should only support the minimum set of features needed for each function. So we should never need to check that a PopRing is in the Closing state because that is impossible.